Panic in zng handling of record alias types

While working on #758 , I noticed that zng also can't handle aliases on records.

```sh
16:04 ~/work/brim/zq(zjson-aliases)
$ cat alias.tzng 
#myrec=record[host:ip]
#1:record[foo:myrec]
1:[[127.0.0.2;]]
16:04 ~/work/brim/zq(zjson-aliases)
$ zq alias.tzng | zq -t -
panic: zng cannot translate type: myrec

goroutine 1 [running]:
github.com/brimsec/zq/zng/resolver.(*Context).TranslateType(0xc000312310, 0x1bb68c0, 0xc0002fd170, 0xc0002cb648, 0x1394158)
	/Users/henridf/work/brim/zq/zng/resolver/context.go:581 +0x2f8
github.com/brimsec/zq/zng/resolver.(*Context).TranslateTypeRecord(0xc000312310, 0xc0002fd1a0, 0x0)
	/Users/henridf/work/brim/zq/zng/resolver/context.go:598 +0xc5
github.com/brimsec/zq/zng/resolver.(*Mapper).Enter(0xc0002e7060, 0x19, 0xc0002fd1a0, 0x0)
	/Users/henridf/work/brim/zq/zng/resolver/mapper.go:26 +0x3d
github.com/brimsec/zq/zio/zngio.(*Reader).Read(0xc0002fd0e0, 0xc0002fd001, 0xc0002fd0e0, 0xc0002cb7b0)
	/Users/henridf/work/brim/zq/zio/zngio/reader.go:83 +0xfa
github.com/brimsec/zq/zio/detector.match(0x1ba20e0, 0xc0002fd0e0, 0x196cf61, 0x3, 0xc0002fd0e0, 0xc0002f3890)
	/Users/henridf/work/brim/zq/zio/detector/reader.go:76 +0x38
github.com/brimsec/zq/zio/detector.NewReaderWithConfig(0x1ba1f20, 0xc0002fcde0, 0xc0002e0540, 0x1975465, 0xa, 0x196d45b, 0x4, 0x0, 0x19a92bd, 0x59, ...)
	/Users/henridf/work/brim/zq/zio/detector/reader.go:56 +0x66f
github.com/brimsec/zq/zio/detector.OpenFromNamedReadCloser(0xc0002e0540, 0x1bab7c0, 0xc0000d6000, 0x1975465, 0xa, 0x196d45b, 0x4, 0x0, 0x19a92bd, 0x59, ...)
	/Users/henridf/work/brim/zq/zio/detector/file.go:149 +0xf0
github.com/brimsec/zq/zio/detector.OpenFile(0xc0002e0540, 0x1975465, 0xa, 0x196d45b, 0x4, 0x0, 0x19a92bd, 0x59, 0x0, 0x1861d20, ...)
	/Users/henridf/work/brim/zq/zio/detector/file.go:74 +0x110
main.(*Command).inputReaders(0xc0001a7180, 0xc0000c01a0, 0x1, 0x1, 0x0, 0x0, 0xd6000000019373e0, 0x1115f01, 0xd668107aa6648996)
	/Users/henridf/work/brim/zq/cmd/zq/zq.go:263 +0x167
main.(*Command).Run(0xc0001a7180, 0xc0000c01a0, 0x1, 0x1, 0x0, 0x0)
	/Users/henridf/work/brim/zq/cmd/zq/zq.go:218 +0x2cf
github.com/mccanne/charm.(*instance).run(0xc0002e6340, 0xc0000c0190, 0x2, 0x2, 0x0, 0x0)
	/Users/henridf/go/pkg/mod/github.com/mccanne/charm@v0.0.3-0.20191224190439-b05e1b7b1be3/instance.go:53 +0x28c
github.com/mccanne/charm.(*Spec).ExecRoot(0x2170960, 0xc0000c0190, 0x2, 0x2, 0xc0002cbf50, 0x100742f, 0xc0000b4058, 0x0)
	/Users/henridf/go/pkg/mod/github.com/mccanne/charm@v0.0.3-0.20191224190439-b05e1b7b1be3/charm.go:77 +0x96
main.main()
	/Users/henridf/work/brim/zq/cmd/zq/main.go:9 +0x78
```

--- a/zbuf/mapper.go
+++ b/zbuf/mapper.go
@@ -28,7 +28,10 @@ func (m *Mapper) Read() (*zng.Record, error) {
 	id := rec.Type.ID()
 	sharedType := m.mapper.Map(id)
 	if sharedType == nil {
-		sharedType = m.mapper.Enter(id, rec.Type)
+		sharedType, err = m.mapper.Enter(id, rec.Type)
+		if err != nil {
+			return nil, err
+		}
 	}
 	rec.Type = sharedType
 	return rec, nil

--- a/zio/zng_test.go
+++ b/zio/zng_test.go
@@ -237,6 +237,9 @@ func TestAlias(t *testing.T) {
 		t.Run("alias-in-different-records", func(t *testing.T) {
 			boomerang(t, multipleRecords)
 		})
+		t.Run("alias-of-record-type", func(t *testing.T) {
+			boomerang(t, recordAlias)
+		})
 	})
 	t.Run("ZJSON", func(t *testing.T) {
 		t.Run("simple", func(t *testing.T) {

--- a/zio/zngio/reader.go
+++ b/zio/zngio/reader.go
@@ -80,12 +80,14 @@ func (r *Reader) Read() (*zng.Record, error) {
 		id := rec.Type.ID()
 		sharedType := r.mapper.Map(id)
 		if sharedType == nil {
-			sharedType = r.mapper.Enter(id, rec.Type)
+			sharedType, err = r.mapper.Enter(id, rec.Type)
+			if err != nil {
+				return nil, err
+			}
 		}
 		rec.Type = sharedType
 		return rec, err
 	}
-
 }
 
 // LastSOS returns the offset of the most recent Start-of-Stream

--- a/zng/resolver/context.go
+++ b/zng/resolver/context.go
@@ -589,6 +589,12 @@ func (c *Context) TranslateType(ext zng.Type) zng.Type {
 		return c.LookupTypeArray(inner)
 	case *zng.TypeUnion:
 		return c.TranslateTypeUnion(ext)
+	case *zng.TypeAlias:
+		typ, err := c.LookupTypeAlias(ext.Name, ext.Type)
+		if err != nil {
+			panic(err)
+		}
+		return typ
 	}
 }
 

--- a/zng/resolver/context.go
+++ b/zng/resolver/context.go
@@ -570,10 +570,10 @@ func (c *Context) parseArrayTypeBody(in string) (string, *zng.TypeArray, error) 
 	return rest, c.LookupTypeArray(inner), nil
 }
 
-func (c *Context) TranslateType(ext zng.Type) zng.Type {
+func (c *Context) TranslateType(ext zng.Type) (zng.Type, error) {
 	id := ext.ID()
 	if id < zng.IdTypeDef {
-		return ext
+		return ext, nil
 	}
 	switch ext := ext.(type) {
 	default:
@@ -582,39 +582,46 @@ func (c *Context) TranslateType(ext zng.Type) zng.Type {
 	case *zng.TypeRecord:
 		return c.TranslateTypeRecord(ext)
 	case *zng.TypeSet:
-		inner := c.TranslateType(ext.InnerType)
-		return c.LookupTypeSet(inner)
+		inner, err := c.TranslateType(ext.InnerType)
+		if err != nil {
+			return nil, err
+		}
+		return c.LookupTypeSet(inner), nil
 	case *zng.TypeArray:
-		inner := c.TranslateType(ext.Type)
-		return c.LookupTypeArray(inner)
+		inner, err := c.TranslateType(ext.Type)
+		if err != nil {
+			return nil, err
+		}
+		return c.LookupTypeArray(inner), nil
 	case *zng.TypeUnion:
 		return c.TranslateTypeUnion(ext)
 	case *zng.TypeAlias:
-		typ, err := c.LookupTypeAlias(ext.Name, ext.Type)
-		if err != nil {
-			panic(err)
-		}
-		return typ
+		return c.LookupTypeAlias(ext.Name, ext.Type)
 	}
 }
 
-func (c *Context) TranslateTypeRecord(ext *zng.TypeRecord) *zng.TypeRecord {
+func (c *Context) TranslateTypeRecord(ext *zng.TypeRecord) (*zng.TypeRecord, error) {
 	var columns []zng.Column
 	for _, col := range ext.Columns {
-		child := c.TranslateType(col.Type)
+		child, err := c.TranslateType(col.Type)
+		if err != nil {
+			return nil, err
+		}
 		columns = append(columns, zng.NewColumn(col.Name, child))
 	}
-
-	return c.MustLookupTypeRecord(columns)
+	return c.MustLookupTypeRecord(columns), nil
 }
 
-func (c *Context) TranslateTypeUnion(ext *zng.TypeUnion) *zng.TypeUnion {
+func (c *Context) TranslateTypeUnion(ext *zng.TypeUnion) (*zng.TypeUnion, error) {
 	var types []zng.Type
 	for _, t := range ext.Types {
-		translated := c.TranslateType(t)
+		translated, err := c.TranslateType(t)
+		if err != nil {
+			return nil, err
+		}
 		types = append(types, translated)
 	}
-	return c.LookupTypeUnion(types)
+	return c.LookupTypeUnion(types), nil
 }
 
 // Cache returns a cache of this table providing lockless lookups, but cannot

--- a/zng/resolver/mapper.go
+++ b/zng/resolver/mapper.go
@@ -22,12 +22,16 @@ func (m *Mapper) Map(td int) *zng.TypeRecord {
 }
 
 //XXX Enter should allocate the td as it creates the new type in the output context
-func (m *Mapper) Enter(id int, ext *zng.TypeRecord) *zng.TypeRecord {
-	if typ := m.outputCtx.TranslateTypeRecord(ext); typ != nil {
-		m.enter(id, typ)
-		return typ
+func (m *Mapper) Enter(id int, ext *zng.TypeRecord) (*zng.TypeRecord, error) {
+	typ, err := m.outputCtx.TranslateTypeRecord(ext)
+	if err != nil {
+		return nil, err
 	}
-	return nil
+	if typ != nil {
+		m.enter(id, typ)
+		return typ, nil
+	}
+	return nil, nil
 }
 
 func (m *Mapper) EnterByName(td int, typeName string) (*zng.TypeRecord, error) {

--- a/zng/resolver/translator.go
+++ b/zng/resolver/translator.go
@@ -21,17 +21,21 @@ func NewTranslator(in, out *Context) *Translator {
 }
 
 // Lookup implements zng.Resolver
-func (t *Translator) Lookup(id int) *zng.TypeRecord {
+func (t *Translator) Lookup(id int) (*zng.TypeRecord, error) {
 	t.mu.Lock()
 	defer t.mu.Unlock()
 	outputType := t.lookup(id)
+	var err error
 	if outputType == nil {
 		inputType := t.inputCtx.Lookup(id)
 		if inputType == nil {
-			return nil
+			return nil, nil
 		}
-		outputType = t.outputCtx.TranslateTypeRecord(inputType)
+		outputType, err = t.outputCtx.TranslateTypeRecord(inputType)
+		if err != nil {
+			return nil, err
+		}
 		t.enter(id, outputType)
 	}
-	return outputType
+	return outputType, nil
 }

